File: 1605.find-valid-matrix-given-row-and-column-sums.cpp
// @leet start
#include <climits>
#include <vector>
using std::vector;

class Solution {
public:
    vector<vector<int>> restoreMatrix(vector<int>& rowSum, vector<int>& colSum) {
        // TC O(mn), SC O(1)

        int numRows = rowSum.size();
        int numCols = colSum.size();

        vector<int> row(numCols); // zero initalisation
        vector<vector<int>> ans(numRows, row);
        // You can also do vec.resize(n) dynamically

        // Fill first column
        for (int i = 0; i < numRows; ++i) {
            ans[i][0] = rowSum[i];
        }

        // Move elements over to next columns
        // numCols - 1 because once the 2nd last col has been moved, guranteed valid
        // j is rowNum
        // i is colNum
        for (int i = 0; i < numCols - 1; ++i) {
            int colTotal = 0;

            for (int j = 0; j < numRows; ++j) {
                colTotal += ans[j][i];
                if (colTotal > colSum[i]) {
                    int diff = std::min(ans[j][i], colTotal - colSum[i]);
                    ans[j][i] -= diff;
                    ans[j][i + 1] = diff;
                    colTotal -= diff;
                }
            } 
        }

        return ans;
                                             
        
    }
};
// @leet end

File: 1605.find-valid-matrix-given-row-and-column-sums.go
package leetcode

// @leet start
// This question specifically doesn't require backtracking, just greedy
// Because any non-negative, can be duplicates

func restoreMatrix(rowSum []int, colSum []int) [][]int {
	numRows := len(rowSum)
	numCols := len(colSum)

	mat := make([][]int, numRows)
	for i := range numRows {
		mat[i] = make([]int, numCols)
	}

	r := 0
	c := 0

	for r < numRows && c < numCols {
		minVar := min(rowSum[r], colSum[c])
		mat[r][c] = minVar

		rowSum[r] -= minVar
		colSum[c] -= minVar

		if rowSum[r] == 0 {
			r++
		} else {
			c++
		}

	}

	return mat

}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// @leet end
